File: backend/git_utils.py
import subprocess
from pathlib import Path
from typing import List, Dict

class GitManager:
    @staticmethod
    def add_worktree(repo_path: str, worktree_path: str, branch: str = None) -> bool:
        cmd = ["git", "-C", repo_path, "worktree", "add", worktree_path]
        if branch:
            cmd.append(branch)
        else:
            cmd.append("-b")
            cmd.append(Path(worktree_path).name)
            
        try:
            subprocess.run(cmd, check=True, capture_output=True)
            return True
        except subprocess.CalledProcessError as e:
            print(f"Error adding worktree: {e.stderr.decode()}")
            return False

    @staticmethod
    def remove_worktree(repo_path: str, worktree_path: str) -> bool:
        try:
            subprocess.run(["git", "-C", repo_path, "worktree", "remove", worktree_path], check=True)
            return True
        except subprocess.CalledProcessError as e:
            print(f"Error removing worktree: {e}")
            return False

    @staticmethod
    def list_worktrees(repo_path: str) -> List[Dict[str, str]]:
        try:
            result = subprocess.run(
                ["git", "-C", repo_path, "worktree", "list", "--porcelain"], 
                check=True, 
                capture_output=True, 
                text=True
            )
            worktrees = []
            current = {}
            for line in result.stdout.splitlines():
                if not line:
                    if current:
                        worktrees.append(current)
                        current = {}
                    continue
                parts = line.split(maxsplit=1)
                if len(parts) == 2:
                    current[parts[0]] = parts[1]
            if current:
                worktrees.append(current)
            return worktrees
        except subprocess.CalledProcessError:
            return []

File: native/src/main/git-utils.ts
import { execSync } from 'child_process';
import { ipcMain } from 'electron';
import path from 'path';
import fs from 'fs';
import os from 'os';

const WORKTREES_ROOT = path.join(os.homedir(), '.overseer', 'worktrees');

if (!fs.existsSync(WORKTREES_ROOT)) {
  fs.mkdirSync(WORKTREES_ROOT, { recursive: true });
}

export const GitManager = {
  addWorktree(repoPath: string, name: string, branch?: string): string | null {
    const worktreePath = path.join(WORKTREES_ROOT, name);
    let cmd = `git -C "${repoPath}" worktree add "${worktreePath}"`;
    if (branch) {
      cmd += ` "${branch}"`;
    } else {
      cmd += ` -b "${name}"`;
    }

    try {
      execSync(cmd);
      return worktreePath;
    } catch (e) {
      console.error('Error adding worktree:', e);
      return null;
    }
  },

  removeWorktree(worktreePath: string): boolean {
    try {
      execSync(`git worktree remove "${worktreePath}" --force`);
      return true;
    } catch (e) {
      console.error('Error removing worktree:', e);
      return false;
    }
  },

  listBranches(repoPath: string): string[] {
    try {
      const output = execSync(`git -C "${repoPath}" branch --format="%(refname:short)"`).toString();
      return output.split('\n').map(b => b.trim()).filter(Boolean);
    } catch (e) {
      console.error('Error listing branches:', e);
      return [];
    }
  }
};

ipcMain.handle('git-add-worktree', async (event, { repoPath, name, branch }) => {
  return GitManager.addWorktree(repoPath, name, branch);
});

ipcMain.handle('git-remove-worktree', async (event, { worktreePath }) => {
  return GitManager.removeWorktree(worktreePath);
});

ipcMain.handle('git-list-branches', async (event, { repoPath }) => {
  return GitManager.listBranches(repoPath);
});
